ERCS/erc-7841.md

```diff
@@ -0,0 +1,545 @@
+
+There is a wide range of ways in which both synchronous and asynchronous protocols may operate. For example, some protocols (particularly for asynchronous messaging) may require the source chain's block in which the message was sent to be finalized (e.g., settled on Ethereum) before it can be included within a block produced for the destination chain. Other protocols may allow for messages to be optimistically included in blocks with consistency checks delayed until settlement time (reminiscent of speculative execution).
+
+### Message Format
+
+All messages MUST follow this format:
+
+```solidity
+/// @title Metadata type
+/// @notice Metadata for a cross-chain message
+struct Metadata {
+    /// @notice The chain identifier of the source chain
+    uint32 srcChainId;
+    /// @notice The chain identifier of the destination chain
+    uint32 destChainId;
+    /// @notice The address of the sending party
+    /// @dev 32 bytes are used to encode the address. In the case of an
+    ///     Ethereum address, the last 12 bytes can be padded with zeros
+    bytes32 srcAddress;
+    /// @notice The address of the recipient
+    /// @dev 32 bytes are used to encode the address. In the case of an 
+    ///     Ethereum address, the last 12 bytes can be padded with zeros
+    bytes32 destAddress;
+    /// @notice The identifier for a cross-chain interaction session
+    /// @dev SHOULD be unique for every new cross-chain calls
+    uint128 sessionId;
+    /// @notice The message counter within an interaction session 
+    /// @dev SHOULD be unique within a session
+    /// @dev OPTIONAL for most asynchronous bridges where every message has a 
+    ///     distinct sessionId, simply set to 0 if unused
+    /// @dev E.g. In a cross-chain call: ChainA.func1 -m1-> ChainB.func2 -m2-> 
+    ///     ChainC.func3 -m3-> ChainB.func4, the subscript i in m_i is the nonce
+    uint128 nonce;
+}
+
+/// @title Message type
+/// @notice A cross-chain message
+struct Message {
+    /// @notice The message metadata 
+    Metadata metadata;
+    /// @notice Message payload encoded using RLP serialization
+    /// @dev It may be ABI-encoded function calls, info about bridged assets, 
+    ///     or arbitrary message data
+    bytes payload;
+}
+```
+
+It is RECOMMENDED that there exists a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID. This work is outside the scope of this ERC, however. To display the sender and receiver of a `Message`, wallets and frontends MAY use [ERC-3770](./eip-3770.md) for chain-specific addresses.
```

Comment:
> It is RECOMMENDED that there exists a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID.

This is a weird way to word this. You're creating a recommendation not for implementers of your standard, but for the entire ecosystem.

I'd recommend something more like:

```suggestion
Implementations SHOULD use a global rollup registry service that supports registration, deregistration, and efficient lookup of a rollup's chain ID. This work is outside the scope of this ERC, however.

To display the sender and receiver of a `Message`, wallets and frontends MAY use [ERC-3770](./eip-3770.md) for chain-specific addresses.
```